src/govuk/i18n.unit.test.mjs

```diff
@@ -39,4 +41,46 @@ describe('I18n', () => {
       expect(() => i18n.t()).toThrow('i18n lookup key missing.')
     })
   })
+
+  describe('string interpolation', () => {
```

Comment:
I think it might be worth adding tests to cover some more of the 'stress cases', things like:

- Replacing placeholders with empty strings, zero or anything else that might be falsey (I think this might fail at the minute)
- Strings that contain things that look a bit like placeholders but aren't, like not having a closing brace
- Nested placeholders like `${name${foo}}`
- Placeholder keys that contain non alphanumeric characters
- Replacements that themselves contain placeholders, i.e. `i18n.t('nameString', { name: '${name}' }) }`
- Placeholders keys that match [function names found on objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object), i.e. `assign`, `create`

Thoughts?

Reply:
I've added tests for the following, as I think it makes sense to 'stress test' a bit more, as you said:

- Replacing placeholders with empty strings, zero or anything else that might be falsey (I think this might fail at the minute)
- Strings that contain things that look a bit like placeholders but aren't, like not having a closing brace
- Placeholder keys that contain non alphanumeric characters
- Placeholders keys that match [function names found on objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object), i.e. assign, create

Let me know if you think those are covered enough now

For the other two (nested placeholders and replacements that themselves contain placeholders), that feels more like additional functionality? I definitely haven't added the logic with that in mind. Is that something we want to support? Related, have you got any examples of how that might be used?

Reply:
Looks great 👍🏻 

For the other two examples, I was thinking purely from the perspective of 'what sorts of inputs could break this?' rather than suggesting additional functionality.

So e.g. checking that we don't somehow cause an infinite loop with a placeholder that replaces with itself, and that the regex behaves 'as expected' when there's a placeholder that looks like `${name${foo}}` (I think it should just replace with a key `name${foo}` from the object?)

Reply:
Ah I see, good point! I'll add some more tests for that.

Reply:
Sorry, that definitely wasn't clear from my original comment!